### DISCUSSION TITLE
Chrome Extension review declined because of "Remote hosted code" (MV3)

### DISCUSSION THREAD

**Discussion by sybers** [InboxSDK/InboxSDK#957] [Q&A] [answered]
Hello,

We have an MV3 Chrome Extension using InboxSDK and our latest deployments on the Chrome Webstore are being rejected because of a violation they call "_Remote hosted code_".

This violation happens when the Chrome Extension loads some scripts which are not contained within the extension code. [This document](https://developer.chrome.com/docs/webstore/program-policies/mv3-requirements/) explains with more details what is expected.

I thought the `@inboxsdk/core` package was supposed to work fine with MV3 apps, since it was part of the code changes when MV3 became mandatory.

Can you confirm if this is related to `@inboxsdk/core`? If yes, will there be a patch soon to prevent that?

If you need more details on our setup I'd be happy to provide them :)

**Comment by wegry** [comment]
> I thought the @inboxsdk/core package was supposed to work fine with MV3 apps

`@inboxsdk/core` should work with MV3 apps. There are some grandfathered in MV2 apps that the SDK still supports.

A few questions. Which version of `@inboxsdk/core`? And what scripts are being loaded that aren't contained within extension code?

**Reply by sybers** [reply]
Hi @wegry,

We're currently using `@inboxsdk/core` version `1.2.1`.

Our Chrome Extension has been using manifest V3 for a while already and was approved in the past...

I was suspecting that this was triggered by [this line](https://github.com/InboxSDK/InboxSDK/blob/e957bfff657e0f7fe2fd3a2b9764738a6e1fd5cd/src/common/load-script.ts#L121).

**Reply by wegry** [reply]
If [the line](https://github.com/InboxSDK/InboxSDK/blob/e957bfff657e0f7fe2fd3a2b9764738a6e1fd5cd/src/common/load-script.ts#L121) in question is what they're balking at, I think we could probably remove it from `@inboxsdk/core` without issue.

Tracing from the gulpfile, we have three output targets. One of which, [inboxsdk-NONREMOTE](https://github.com/InboxSDK/InboxSDK/blob/e957bfff657e0f7fe2fd3a2b9764738a6e1fd5cd/gulpfile.babel.ts#L386, transitively imports `load-script` via this line https://github.com/InboxSDK/InboxSDK/blob/e957bfff657e0f7fe2fd3a2b9764738a6e1fd5cd/src/inboxsdk-js/inboxsdk-NONREMOTE.ts#L2, only to monkey patch it at runtime [here](https://github.com/InboxSDK/InboxSDK/blob/e957bfff657e0f7fe2fd3a2b9764738a6e1fd5cd/src/inboxsdk-js/inboxsdk-NONREMOTE.ts#L4).

One wrinkle is that we depend on `load-script` at Streak with the remote build on the SDK. 

I'll check in with folks in American Pacific Time to see what our options are.

**Reply by sybers** [reply]
Thanks a lot @wegry!

Looking forward to see if this could be fixed.

**Reply by wegry** [reply]
Before we try our hands at a fix, @sybers are you sure the Chrome Web Store rejection is happening because of the `@inboxsdk/core` code being included?

**Reply by sybers** [reply]
I'm pretty confident it is, but I cannot be 100% sure since they remain vague when we ask for more precision about what exactly caused the rejection.

I have spotted another dependency on our side which has some `eval` like code. And even though it's dead code I would like to try removing it to see if they accept the submission.

I'll let you know here if it is still rejected after removing this dependency as I understand you don't want to "fix" the above thing if in the end it's not the root cause of my troubles!

**Reply by wegry** [reply]
🙏 Sounds good.

**Reply by sybers** [reply]
Hello @wegry,

Against all odds, our Chrome Extension was published without having to do any code change... 🤦‍♂️

It looks like they report false positives or their reviewing system is flawed, but at least it means no code changes required!

If this happens again I'll come back to this discussion, otherwise thanks a lot for your time :)

**Reply by sybers** [reply]
Hi @wegry, our extension has been rejected once again, and no code was changed on our side...

I think they complain about loading `https://www.inboxsdk.com/build/platform-implementation.js` from a remote server.

Basically any `.js` file loaded from an extenal source would be problematic. Can InboxSDK do anything about that? I'm surprised our extension is rejected because of that, when many others also using InboxSDK are accepted 🤔

**Reply by wegry** [reply]
We do have a little preliminary work around removing that loading code from the npm build, @Macil's #970.

Just to confirm though, previously you'd mentioned

> I have spotted another dependency on our side which has some eval like code. And even though it's dead code I would like to try removing it to see if they accept the submission.

Which dependency, if you can disclose, and are you still using that as well?

**Reply by aleemstreak** [reply]
@sybers did the rejection mention anything specific about the inboxsdk?

**Reply by sybers** [reply]
@aleemstreak unfortunately not, they only mention the file in which they found the issue. That's all we can know 😐

**Reply by wegry** [reply, marked as answer]
@sybers we cut @inboxsdk/core@2.0.0 recently. Let us know if it helps with your Chrome Web Store submission.

**Reply by Decafeinato** [reply]
Hi, I got the same rejection today. I'm working with the  version 1.1.8 of inboxsdk/core package.
I'm going to make an attempt to upgrade to 2.++ . 

@sybers  Do you have any update on your rejection case ?

David

**Reply by wegry** [reply]
@Decafeinato we did have another user reach out via email. They've had an extension that was rejected accepted after updating to `@inboxsdk/core@^2.0.0`.

**Reply by Decafeinato** [reply]
Thanks a lot! I will keep you posted👍

**Reply by sybers** [reply]
@Decafeinato Our extension was rejected after 2.0, then accepted again, then rejected again... The review process on Google's side is strange. But now I can say that Inboxsdk doesn't seem to be causing these rejections anymore :)

**Reply by Decafeinato** [reply]
Thanks a lot! Yes I'm surprised too there are now rejecting me on some stuff that have been there since day 1 ... and caused by their own packages😅.

**Reply by Decafeinato** [reply]
I confirm that the bump to 2.x.x allows me to publish again

**Comment by aleemstreak** [comment]
I wonder if you could split up your files so that the SDK is only
referenced in one of them so we can be sure that that is the issue.

**Comment by aleemstreak** [comment]
If you've experienced rejections using the latest NPM build would you mind
emailing me directly at aleem at streak dot com with your extension id? I'm
going to follow up with the google team to see if they can shed any light
on it. You're correct though, it may have nothing to do with the InboxSDK.
ᐧ

**Reply by sybers** [reply]
Hey @aleemstreak if you don't mind I'd love to send you our extension ID as well. If it can help us grab more information on why it's being rejected, even if it's not related to Inboxsdk, it would definitely help us!

**Comment by aleemstreak** [comment]
Yeah they won’t tell us why an extension is rejected but they might share
if it has something to do with the SDK
